gateway/gateway.go

```diff
@@ -1670,3 +1632,112 @@ func WithContentType(contentType string, delegate http.HandlerFunc) http.Handler
 		delegate(w, r)
 	})
 }
+
+func injectIdentifiers(userIDHeader string, body []byte) (res []byte, notIdentifiable, nonRudderEvent, containsAudienceList bool, builtUserID string) {
+	// set anonymousId if not set in payload
+	result := gjson.GetBytes(body, "batch")
+	var out []map[string]interface{}
+	result.ForEach(func(_, vjson gjson.Result) bool {
+		anonIDFromReq := strings.TrimSpace(vjson.Get("anonymousId").String())
+		userIDFromReq := strings.TrimSpace(vjson.Get("userId").String())
+		if builtUserID == "" {
+			if anonIDFromReq != "" {
+				builtUserID = userIDHeader + DELIMITER + anonIDFromReq + DELIMITER + userIDFromReq
+			} else {
+				// Proxy gets the userID from body if there is no anonymousId in body/header
+				builtUserID = userIDHeader + DELIMITER + userIDFromReq + DELIMITER + userIDFromReq
+			}
+		}
+
+		eventTypeFromReq := strings.TrimSpace(vjson.Get("type").String())
+
+		if anonIDFromReq == "" {
+			if userIDFromReq == "" && !allowReqsWithoutUserIDAndAnonymousID {
+				notIdentifiable = true
+				return false
+			}
+		}
+		if eventTypeFromReq == "audiencelist" {
+			containsAudienceList = true
+		}
+		// hashing combination of userIDFromReq + anonIDFromReq, using colon as a delimiter
+		rudderId, err := misc.GetMD5UUID(userIDFromReq + ":" + anonIDFromReq)
+		if err != nil {
+			notIdentifiable = true
+			return false
+		}
+
+		toSet, ok := vjson.Value().(map[string]interface{})
+		if !ok {
+			nonRudderEvent = true
+			return false
+		}
+		toSet["rudderId"] = rudderId
+		if messageId := strings.TrimSpace(vjson.Get("messageId").String()); messageId == "" {
+			toSet["messageId"] = uuid.Must(uuid.NewV4()).String()
+		}
+		out = append(out, toSet)
+		return true // keep iterating
+	})
+
+	res, _ = sjson.SetBytes(body, "batch", out)
+	return
+}
+
+func injectIdentifiersV2(userIDHeader string, body []byte) (res []byte, notIdentifiable, nonRudderEvent, containsAudienceList bool, builtUserID string) {
+	type batchMessage struct {
+		Batch []map[string]interface{} `json:"batch"`
```

Comment:
We can also test:
```suggestion
        type usefulFields struct {
            AnonymousID string 
            UserID string
            Type string
        }
		Batch []usefulFields `json:"batch"`
```
It is more memory efficient for sure, but maybe faster for the CPU as well

Reply:
Ignore, for now, I will try to give a proposal that works end 2 end

Reply:
I was quick to comment without reading the rest of the function:

here is my suggestion:
https://github.com/rudderlabs/rudder-server/pull/2506/files#diff-17cb8b37eda9018fe1c6cdb5f96b3fc948fc8ba49bc516987b8269576db9fcd4R1812

Reply:
v4 looks nice :)